README.md
Gallery - plg_content_gallery
=============

Plugin Joomla para exibição de galerias de imagens a partir de uma pasta.

Requirements
------------

* [Joomla! 2.5+](http://www.joomla.org)

Author
------

Rene Bentes Pinto

License
--------

* This Plugin licensed under the terms of the [GNU/GPLv2](http://www.gnu.org/licenses/gpl-2.0.html) license.

Bugs/Requests
-------------

You can [report a bug or request a feature here](http://github.com/renebentes/plg_content_gallery/issues)

TODO
----



Release Notes
-------------

You can [see the release notes here](http://github.com/renebentes/plg_content_gallery/blob/master/CHANGELOG.md)

CHANGELOG.md
CHANGELOG
=============

Este arquivo refere-se às mudanças realizadas desde a versão inicial de desenvolvimento.

* 1.0.5 - 09/09/2014
  * Fixed - Bugs corrigidos
  * Added - Método OnContentBeforeDysplay para incluir css e js adicional
* 1.0.4 - 26/08/2014
  * Removed - Referência às alterações sofridas pelo projeto removidas do arquivo README.md
  * Added - Arquivo CHANGELOG.md com as alterações sofridas pelo projeto
* 1.0.3 - 22/08/2014
  * Added - CSS para botão fechar modal
  * Updated - Caption para imagens
  * Added - Título para modal
* 1.0.2 - 15/08/2014
  * Added - Script de instalação para remoção de arquivos antigos
  * Fixed - Corrigido referência para JFile
* 1.0.1 - 13/08/2014
  * Updated - Novo tamanho para thumbnails
* 1.0.0 - 04/08/2014
  * Message - Estágio inicial de desenvolvimento

helper.php
<?php
/**
 * @package     Gallery
 * @subpackage  plg_gallery
 * @copyright   Copyright (C) 2014 Rene Bentes Pinto, Inc. All rights reserved.
 * @license     GNU General Public License version 2 or later; see http://www.gnu.org/licenses/gpl-2.0.html
 */

// No direct access.
defined('_JEXEC') or die('Restricted access!');

jimport('joomla.filesystem.folder');
jimport('joomla.filesystem.file');

/**
* Gallery plugin helper class
*
* @package    Gallery
* @subpackage plg_gallery
* @since      2.5
*/
abstract class GalleryHelper
{
  /**
   * Get items to Gallery
   *
   * @param  string  $folder      Path to image files
   * @param  string  $size        Thumbnail size (width x height)
   * @param  boolean $smartResize Type of resize
   * @param  integer $quality     JPEG quality
   * @param  integer $expireTime  Cache time
   * @param  integer $id          Gallery identifier
   *
   * @return ob                Gallery items
   */
  public static function getGallery($folder, $size = '200x160', $smartResize = true, $quality = 90, $expireTime = 86400, $id)
  {
    $folder  = JPATH_SITE . '/images/' . $folder;

    return self::_createThumbnail($folder, $size, $smartResize, $quality, $expireTime, $id);
  }

  /**
   * Gets captions for items
   *
   * @param string $folder    The gallery path
   * @param string $reference The item gallery filename
   *
   * @return object           Captions
   */
  private static function _getCaption($folder, $reference)
  {
    $file                 = $folder . '/gallery.txt';
    $caption              = new JObject;
    $caption->title       = '';
    $caption->description = '';

    if (JFile::exists($file))
    {
      $content = file($file);

      foreach ($content as $line)
      {
        $temp = explode(';', $line);
        if (strtolower($temp[0]) == strtolower($reference))
        {
          $caption->title       = nl2br($temp[1]);
          $caption->description = nl2br($temp[2]);
          break;
        }
      }
    }
    return $caption;
  }

  /**
   * Get the path to a layout from Gallery
   *
   * @param   string  $type    Plugin type
   * @param   string  $name    Plugin name
   * @param   string  $layout  Layout name
   *
   * @return  string  Layout path
   *
   * @since   2.5
   */
  public static function getLayoutPath($type, $name, $layout = 'default')
  {
    $template      = JFactory::getApplication()->getTemplate();
    $defaultLayout = $layout;

    if (strpos($layout, ':') !== false)
    {
      // Get the template and file name from the string
      $temp          = explode(':', $layout);
      $template      = ($temp[0] == '_') ? $template : $temp[0];
      $layout        = $temp[1];
      $defaultLayout = ($temp[1]) ? $temp[1] : 'default';
    }

    // Build the template and base path for the layout
    $tPath = JPATH_THEMES . '/' . $template . '/html/plg_' . $type . '_' . $name . '/' . $layout . '.php';
    $bPath = JPATH_BASE . '/plugins/' . $type . '/' . $name . '/tmpl/' . $defaultLayout . '.php';
    $dPath = JPATH_BASE . '/plugins/' . $type . '/' . $name . '/tmpl/default.php';

    // If the template has a layout override use it
    if (file_exists($tPath))
    {
      return $tPath;
    }
    elseif (file_exists($bPath))
    {
      return $bPath;
    }
    else
    {
      return $dPath;
    }
  }

  /**
   * Creates thumbnails for images
   *
   * @param  string  $folder      Path to image files
   * @param  string  $size        Thumbnail size (width x height)
   * @param  boolean $smartResize Type of resize
   * @param  integer $quality     JPEG quality
   * @param  integer $expireTime  Cache time
   * @param  integer $id          Gallery identifier
   *
   * @return array                Gallery thumbnails
   */
  private static function _createThumbnail($folder, $size, $smartResize, $quality, $expireTime, $id)
  {
    if (!JFolder::exists($folder))
    {
      JFactory::getApplication()->enqueueMessage(JText::_('PLG_CONTENT_GALLERY_ERROR_PATH_NOT_FOUND'), 'error');
      return;
    }

    $files = JFolder::files($folder);

    if (!$files)
    {
      JFactory::getApplication()->enqueueMessage(JText::_('PLG_CONTENT_GALLERY_ERROR_FILES_NOT_FOUND'), 'error');
      return;
    }

    $size = explode('x', strtolower($size));

    if (count($size) != 2)
    {
      JFactory::getApplication()->enqueueMessage(JText::sprintf('PLG_CONTENT_GALLERY_ERROR_ARGUMENT_SIZE_INVALID', $size), 'error');
      return;
    }

    if (!is_numeric($size[0]) || $size[0] < 0)
    {
      JFactory::getApplication()->enqueueMessage(JText::sprintf('PLG_CONTENT_GALLERY_ERROR_WIDTH_INVALID', $size[0]), 'error');
      return;
    }

    if (!is_numeric($size[1]) || $size[1] < 0)
    {
      JFactory::getApplication()->enqueueMessage(JText::sprintf('PLG_CONTENT_GALLERY_ERROR_HEIGHT_INVALID', $size[1]), 'error');
      return;
    }

    $cache = JPATH_SITE . '/cache/plg_content_gallery/' . $id;

    if (JFolder::exists($cache) || JFolder::create($cache))
    {
      $fileTypes = array('jpg', 'jpeg', 'gif', 'png');
      $found     = array();

      foreach ($files as $file)
      {
        $fileInfo = pathinfo($file);
        if (array_key_exists('extension', $fileInfo) && in_array(strtolower($fileInfo['extension']), $fileTypes))
        {
          $found[] = $file;
        }
      }

      // Bail out if there are no images found
      if (count($found) == 0)
      {
        JFactory::getApplication()->enqueueMessage(JText::_('PLG_CONTENT_GALLERY_ERROR_IMAGES_NOT_FOUND'), 'error');
        return;
      }

      sort($found);
      $gallery = array();

      foreach ($found as $key => $filename)
      {
        $gallery[$key] = new JObject;

        if (self::_getCaption($folder, $filename))
        {
          $gallery[$key]->title       = self::_getCaption($folder, $filename)->title;
          $gallery[$key]->description = self::_getCaption($folder, $filename)->description;
        }

        $original  = $folder . '/' . $filename;
        $thumbname = $cache . '/' . strtolower($filename);

        // Check if thumbnail exists already
        if (!(JFile::exists($thumbname) && is_readable($thumbname) && (filemtime($thumbname) + $expireTime) > time()))
        {
          list($width, $height, $type) = getimagesize($original);

          switch($type)
          {
            case 1 :
              $source = @ imagecreatefromgif($original);
              break;
            case 2 :
              $source = imagecreatefromjpeg($original);
              break;
            case 3 :
              $source = imagecreatefrompng($original);
              break;
            default :
              $source = NULL;
          }

          if (!$source)
          {
            JFactory::getApplication()->enqueueMessage(JText::_('PLG_CONTENT_GALLERY_ERROR_PROCESS_SOURCE'), 'error');
            return;
          }

          $thumbnail = self::_getDimension($width, $height, $size[0], $size[1], $smartResize);

          $thumb = imagecreatetruecolor($size[0], $size[1]);
          imagecopyresampled($thumb, $source, 0, 0, 0, 0, $size[0], $size[1], $width, $height);

          switch($type)
          {
            case 1 :
              $success = imagegif($thumb, $thumbname);
              break;
            case 2 :
              $success = imagejpeg($thumb, $thumbname, $quality);
              break;
            case 3 :
              $success = imagepng($thumb, $thumbname);
              break;
            default :
              $success = NULL;
          }

          // Bail out if there is a problem in the GD conversion
          if (!$success)
          {
            return;
          }

          // remove the image resources from memory
          imagedestroy($source);
          imagedestroy($thumb);
        }

        $gallery[$key]->filename  = $filename;
        $gallery[$key]->image     = str_replace(JPATH_SITE, JUri::base(true), $original);
        $gallery[$key]->thumbnail = str_replace(JPATH_SITE, JUri::base(true), $thumbname);
      }
    }

    return $gallery;
  }

  /**
   * Calculate the dimensions of the thumbnails
   *
   * @param  integer $width       Source width
   * @param  integer $height      Source height
   * @param  integer $thb_width   Thumbnail width
   * @param  integer $thb_height  Thumbnail height
   * @param  boolean $smartResize Resize type
   *
   * @return array                The thumbnail dimensions
   */
  private static function _getDimension($width, $height, $thb_width, $thb_height, $smartResize)
  {
    if ($smartResize)
    {
      // thumb ratio bigger that container ratio
      if ($width / $height > $thb_width / $thb_height)
      {
        // wide containers
        if ($thb_width >= $thb_height)
        {
          // wide thumbs
          if ($width > $height)
          {
            $thumb_width  = $thb_height * $width / $height;
            $thumb_height = $thb_height;
          }
          // high thumbs
          else
          {
            $thumb_width  = $thb_height * $width / $height;
            $thumb_height = $thb_height;
          }
          // high containers
        }
        else
        {
          // wide thumbs
          if ($width > $height)
          {
            $thumb_width  = $thb_height * $width / $height;
            $thumb_height = $thb_height;
          }
          // high thumbs
          else
          {
            $thumb_width  = $thb_height * $width / $height;
            $thumb_height = $thb_height;
          }
        }
      }
      else
      {
        // wide containers
        if ($thb_width >= $thb_height)
        {
          // wide thumbs
          if ($width > $height)
          {
            $thumb_width  = $thb_width;
            $thumb_height = $thb_width * $height / $width;
          }
          // high thumbs
          else
          {
            $thumb_width  = $thb_width;
            $thumb_height = $thb_width * $height / $width;
          }
          // high containers
        }
        else
        {
          // wide thumbs
          if ($width > $height)
          {
            $thumb_width  = $thb_height * $width / $height;
            $thumb_height = $thb_height;
          }
          // high thumbs
          else
          {
            $thumb_width  = $thb_width;
            $thumb_height = $thb_width * $height / $width;
          }
        }
      }
    }
    else
    {
      if ($width > $height)
      {
        $thumb_width  = $thb_width;
        $thumb_height = $thb_width * $height / $width;
      }
      elseif ($width < $height)
      {
        $thumb_width  = $thb_height * $width / $height;
        $thumb_height = $thb_height;
      }
      else
      {
        $thumb_width  = $thb_width;
        $thumb_height = $thb_height;
      }
    }

    $thumbnail           = array();
    $thumbnail['width']  = round($thumb_width);
    $thumbnail['height'] = round($thumb_height);

    return $thumbnail;

  }
}

script.php
<?php
/**
 * @package     Gallery
 * @subpackage  plg_content_galler
 * @copyright   Copyright (C) 2014 Rene Bentes Pinto, Inc. All rights reserved.
 * @license     GNU General Public License version 2 or later; see http://www.gnu.org/licenses/gpl-2.0.html
 */

// No direct access.
defined('_JEXEC') or die('Restricted access!');

// Include dependencies
jimport('joomla.filesystem.file');
jimport('joomla.filesystem.folder');

/**
 * Script file of Gallery Plugin
 *
 * @package     Gallery
 * @subpackage  plg_content_gallery
 *
 * @since       2.5
 */
class PlgContentGalleryInstallerScript
{
  /**
   * Extension name
   *
   * @var string
   */
  private $_extension = 'plg_content_gallery';

  /**
   * Version release
   *
   * @vare string
   */
  private $_release = '';

  /**
   * Array of sub extensions package
   *
   * @var array
   */
  private $_subextensions = array(
    'modules' => array(
    ),
    'plugins' => array(
    )
  );

  /**
   * Array of obsoletes files and folders
   *
   * @var array
   */
  private $_obsoletes = array(
    'files' => array(
      'plugins/content/gallery/assets/gallery.css',
      'plugins/content/gallery/assets/js/blueimp-gallery.js',
      'plugins/content/gallery/assets/js/blueimp-gallery.min.js',
    ),
    'folders' => array(
    )
  );

  /**
   * Method to install the plugin
   *
   * @param JInstaller $parent
   */
  function install($parent)
  {
    echo '<p>' . JText::sprintf('PLG_CONTENT_GALLERY_INSTALL_TEXT', $this->_extension, $this->_release) . '</p>';
  }

  /**
   * Method to uninstall the plugin
   *
   * @param JInstaller $parent
   */
  function uninstall($parent)
  {
    echo '<p>' . JText::sprintf('PLG_CONTENT_GALLERY_UNINSTALL_TEXT', $this->_extension, $this->_release) . '</p>';
  }

  /**
   * Method to update the plugin
   *
   * @param JInstaller $parent
   */
  function update($parent)
  {
    echo '<p>' . JText::sprintf('PLG_CONTENT_GALLERY_UPDATE_TEXT', $this->_extension, $this->_release) . '</p>';
  }

  /**
   * Method to run before an install/update/uninstall method
   *
   * @param string     $type Installation type (install, update, discover_install)
   * @param JInstaller $parent Parent object
   */
  function preflight($type, $parent)
  {
    $this->_checkCompatible($type, $parent);
  }

  /**
   * Method to run after an install/update/uninstall method
   *
   * @param string     $type install, update or discover_update
   * @param JInstaller $parent
   */
  function postflight($type, $parent)
  {
    $this->_removeObsoletes($this->_obsoletes);
  }

  /**
   * Method for checking compatibility installation environment
   *
   * @param  JInstaller   $parent Parent object
   *
   * @return bool         True if the installation environment is compatible
   */
  private function _checkCompatible($type, $parent)
  {
    // Get the application.
    $app            = JFactory::getApplication();
    $this->_release = (string) $parent->get('manifest')->version;
    $min_version    = (string) $parent->get('manifest')->attributes()->version;
    $jversion       = new JVersion;

    if (version_compare($jversion->getShortVersion(), $min_version, 'lt' ))
    {
      $app->enqueueMessage(JText::sprintf('PLG_CONTENT_GALLERY_VERSION_UNSUPPORTED', $this->_extension, $this->_release, $min_version), 'error');
      return false;
    }

    // Storing old release number for process in postflight.
    if ($type == 'update')
    {
      $oldRelease = $this->_getParam('version');

      if (version_compare($this->_release, $oldRelease, 'lt'))
      {
        $app->enqueueMessage(JText::sprintf('PLG_CONTENT_GALLERY_UPDATE_UNSUPPORTED', $this->_extension, $oldRelease, $this->_release), 'error');
        return false;
      }
    }
  }

  /**
   * Removes obsolete files and folders
   *
   * @param array $obsoletes
   */
  private function _removeObsoletes($obsoletes = array())
  {
    // Remove files
    if(!empty($obsoletes['files']))
    {
      foreach($obsoletes['files'] as $file)
      {
        $f = JPATH_ROOT . '/' . $file;
        if(!JFile::exists($f))
        {
          continue;
        }
        JFile::delete($f);
      }
    }

    // Remove folders
    if(!empty($obsoletes['folders']))
    {
      foreach($obsoletes['folders'] as $folder)
      {
        $f = JPATH_ROOT . '/' . $folder;
        if(!JFolder::exists($f))
        {
          continue;
        }
        JFolder::delete($f);
      }
    }
  }

  /**
   * Get a variable from the manifest cache.
   *
   * @param string $name Column name
   *
   * @return string Value of column name
   */
  private function _getParam($name)
  {
    $db    = JFactory::getDbo();
    $query = $db->getQuery(true);

    $query->select($db->quoteName('manifest_cache'));
    $query->from($db->quoteName('#__extensions'));
    $query->where($db->quoteName('name') . ' = ' . $db->quote($this->_extension));
    $db->setQuery($query);

    $manifest = json_decode($db->loadResult(), true);

    return $manifest[$name];
  }
}

tmpl/default.php
<?php
/**
 * @package     Gallery
 * @subpackage  plg_gallery
 * @copyright   Copyright (C) 2014 Rene Bentes Pinto, Inc. All rights reserved.
 * @license     GNU General Public License version 2 or later; see http://www.gnu.org/licenses/gpl-2.0.html
 */

// No direct access.
defined('_JEXEC') or die('Restricted access!');

?>

<div class="row gallery" data-title="<?php echo $row->title; ?>">
<?php foreach ($row->gallery as $key => $photo) : ?>
  <div class="col-md-3">
    <a href="<?php echo $photo->image; ?>" class="thumbnail" data-toggle="gallery" data-title="<?php echo $photo->title; ?>" data-description="<?php echo $photo->description; ?>">
      <img class="img-gallery img-responsive" src="<?php echo $photo->thumbnail; ?>" alt="<?php echo $photo->title;?>" title="<?php echo $photo->title;?>" />
    </a>
  </div>
<?php endforeach; ?>
</div>

assets/js/gallery.js
/*!
 * Gallery v1.0.0
 * Copyright 2014 Rene Bentes Pinto
 * Licensed under GNU General Public License version 2 or later; see http://www.gnu.org/licenses/gpl-2.0.html
 */

if (typeof jQuery === 'undefined') throw new Error('Gallery\'s Javascript requires jQuery!');

+ function($) {
  'use strict';

  // GALLERY CLASS DEFINITION
  // ========================

  var Gallery = function(element, options) {
    this.$body     = $(document.body)
    this.$element  = $(element);
    this.options   = $.extend({}, Gallery.DEFAULTS, options);
    this.$parent   = this.$element.parents().find('.gallery');
    this.count     = this.$parent.children().length;
    this.index     = this.$element.parent().index();
    this.container = this.options.container;

    this.template = {
      dialog: '<div class="gallery-modal modal fade" tabindex="-1" role="dialog" aria-labelledby="modalLabel" aria-hidden="true">' +
        '<div class="modal-dialog">' +
        '<div class="modal-content">' +
        '<div class="modal-body">' +
        '</div>' +
        '</div>' +
        '</div>' +
        '</div>',
      header: '<div class="modal-header">' +
        '<button type="button" class="btn btn-default btn-sm" data-dismiss="modal">' +
        '<span class="fa fa-times" aria-hidden="true"></span>' +
        '<span class="sr-only">Close</span>' +
        '</button>' +
        '</div>',
      image: '<img src="" class="img-responsive center-block" alt="" title="" />',
      footer: '<div class="modal-footer"></div>',
      legend: '<div class="caption"></div>',
      control: '<div class="btn-group control">' +
        '<button type="button" class="btn btn-default prev"><i class="fa fa-chevron-left"></i></button>' +
        '<button type="button" class="btn btn-default next"><i class="fa fa-chevron-right"></i></button>' +
        '</div>'
    };
  };

  Gallery.VERSION = '1.0.3';

  Gallery.DEFAULTS = {
    container: '.gallery-modal'
  };

  Gallery.prototype.show = function() {
    $('div').remove(this.container);
    this.$body.append(this.template.dialog);

    var galleryTitle = this.$parent.attr('data-title');
    var title = this.$element.attr('data-title');
    var description = this.$element.attr('data-description');

    $(this.container).on('keydown.bs.gallery', $.proxy(this.keydown, this));

    $(this.container + ' .modal-content').prepend(this.template.header);

    if (galleryTitle !== 'undefined')
      $(this.container + ' .modal-header').append('<h4 class="modal-title">' + this.$parent.data('title') + '</h4>');

    $(this.container + ' .modal-body').append(this.template.image);
    $(this.container + ' img').attr('src', this.$element.attr('href'));


    if (this.count > 1 || (title !== 'undefined' || description !== 'undefined')) {
      $(this.container + ' .modal-content').append(this.template.footer);

      if (title !== 'undefined' || description !== 'undefined') {
        $(this.container + ' .modal-body').append(this.template.legend);

        if (title !== 'undefined')
          $(this.container + ' .modal-body .caption').append('<h4>' + title + '</h4>');

        if (description !== 'undefined')
          $(this.container + ' .modal-body .caption').append('<p>' + description + '</p>');
      }

      if (this.count > 1) {
        $(this.container + ' .modal-footer').append(this.template.control);
        $(this.container + ' .control .prev').on('click.prev.bs.gallery', $.proxy(this.prev, this));
        $(this.container + ' .control .next').on('click.next.bs.gallery', $.proxy(this.next, this));
      }
    }

    $(this.container).modal();
  };

  Gallery.prototype.keydown = function(e) {
    switch (e.which) {
      case 37:
        this.prev();
        break;
      case 39:
        this.next();
        break;
      case 27:
        $(this.container).modal('hide');
        break;
      default:
        return;
    }
    e.preventDefault();
  };

  Gallery.prototype.prev = function() {
    this.index--;
    $('div').remove('.modal-body .caption');
    if (this.index < 0) {
      this.index = this.count - 1;
    }

    $(this.container + ' img').attr('src', this.$parent.find('a').get(this.index).getAttribute('href'));
    var title = this.$parent.find('a').get(this.index).getAttribute('data-title');
    var description = this.$parent.find('a').get(this.index).getAttribute('data-description');

    if (title !== 'undefined' || description !== 'undefined') {
      $(this.container + ' .modal-body').append(this.template.legend);
      $(this.container + ' .caption').append('<h4>' + title + '</h4>');
      $(this.container + ' .caption').append('<p>' + description + '</p>');
    }

    return false;
  };

  Gallery.prototype.next = function() {
    this.index++;
    $('div').remove('.modal-body .caption');
    if (this.index >= this.count) {
      this.index = 0;
    }

    $(this.container + ' img').attr('src', this.$parent.find('a').get(this.index).getAttribute('href'));
    var title = this.$parent.find('a').get(this.index).getAttribute('data-title');
    var description = this.$parent.find('a').get(this.index).getAttribute('data-description');

    if (title !== 'undefined' || description !== 'undefined') {
      $(this.container + ' .modal-body').append(this.template.legend);
      $(this.container + ' .caption').append('<h4>' + title + '</h4>');
      $(this.container + ' .caption').append('<p>' + description + '</p>');
    }

    return false;
  };

  // GALLERY PLUGIN DEFINITION
  // ==================================

  function Plugin(option) {
    return this.each(function() {
      var $this = $(this);
      var data = $this.data('bs.gallery');
      var options = typeof option == 'object' && option;

      if (!data) $(this).data('bs.gallery', (data = new Gallery(this, options)));
      if (typeof option == 'string') data[option]();
    });
  };

  var old = $.fn.gallery;

  $.fn.gallery = Plugin;
  $.fn.gallery.Constructor = Gallery;

  // GALLERY NOCONFLICT

  $.fn.gallery.noConflict = function() {
    $.fn.gallery = old;
    return this;
  }

  // GALLERY DATA-API
  // ==============

  $(document).on('click.bs.gallery.data-api', '[data-toggle="gallery"]', function(e) {
    e.preventDefault();

    Plugin.call($(this), 'show');
  });

}(jQuery);

gallery.php
<?php
/**
 * @package     Gallery
 * @subpackage  plg_gallery
 * @copyright   Copyright (C) 2014 Rene Bentes Pinto, Inc. All rights reserved.
 * @license     GNU General Public License version 2 or later; see http://www.gnu.org/licenses/gpl-2.0.html
 */

// No direct access.
defined('_JEXEC') or die('Restricted access!');

/**
 * Gallery Content plugin
 *
 * @package     Gallery
 * @subpackage  plg_content_gallery
 * @since       2.5
 */
class plgContentGallery extends JPlugin
{
  /**
   * Affects constructor behavior. If true, language files will be loaded automatically.
   *
   * @var    boolean
   * @since  3.1
   */
  protected $autoloadLanguage = true;

  /**
   * Constructor
   *
   * @param   object  &$subject  The object to observe
   * @param   array   $config    An array that holds the plugin configuration
   *
   * @since   2.5
   */
  public function __construct(&$subject, $config)
  {
    parent::__construct($subject, $config);

    // Load the language files if needed.
    if ($this->autoloadLanguage)
    {
      $this->loadLanguage();
    }
  }

  /**
   * Method called to prepare content before html output
   *
   * @param string  $context    The context of the content being passed to the plugin.
   * @param object  &$item      The content object.  Note $item->text is also available
   * @param object  &$params    The content params.
   * @param integer $page       The 'page' number.
   *
   * @return mixed Returns void on success or false otherwise
   *
   * @since 2.5
   */
  public function onContentPrepare($context, &$row, &$params, $page = 0)
  {
    require_once dirname(__FILE__) . '/helper.php';

    // Simple performance check to determine whether bot should process further.
    if (JString::strpos($row->text, '{gallery}') === false)
    {
      return;
    }

    $regex = "/{gallery}(.+?){\/gallery}/is";
    preg_match_all($regex, $row->text, $matches);

    if (!count($matches))
    {
      return;
    }

    // Get the path for the layout file
    $version = new JVersion;
    if ($version->isCompatible(3.0)) {
      $path = JPluginHelper::getLayoutPath('content', 'gallery');
    }
    else {
      $path = GalleryHelper::getLayoutPath('content', 'gallery');
    }

    foreach ($matches[1] as $source)
    {
      $row->gallery = GalleryHelper::getGallery($source,'200x160', true, 90, 86400, $row->id);
      ob_start();
      include $path;
      $html = ob_get_contents();
      ob_end_clean();

      $regex = '/{gallery}' . str_replace('.', '\.', str_replace('/', '\/', $source)) . '{\/gallery}/is';

      $row->text = preg_replace($regex, $html, $row->text);
    }


    return true;
  }

  /**
   * Method is called by the view and the results are imploded and displayed in a placeholder
   *
   * @param string  $context    The context of the content passed to the plugin.
   * @param object  &$item      The content object.  Note $item->text is also available
   * @param object  &$params    The content params
   * @param integer $page       The 'page' number
   *
   * @return string
   *
   * @since 2.5
   */
  public function onContentBeforeDisplay($context, &$item, &$params, $page = 0)
  {
    JFactory::getDocument()->addStylesheet(JUri::root(true) . '/plugins/content/gallery/assets/css/gallery.css');
    echo '<script src="' . JUri::root(true) . '/plugins/content/gallery/assets/js/gallery.js" type="text/javascript"></script>';

    return '';
  }
}